cmd/deploy_test.go

```diff
@@ -1,182 +1,421 @@
+		t.Fatal(err)
+	}
+	if f.Git.URL != url {
+		t.Errorf("expected git URL '%v' got '%v'", url, f.Git.URL)
+	}
+	if f.Git.Revision != branch {
+		t.Errorf("expected git branch '%v' got '%v'", branch, f.Git.Revision)
+	}
+	if f.Git.ContextDir != dir {
+		t.Errorf("expected git dir '%v' got '%v'", dir, f.Git.ContextDir)
+	}
+}
+
+// TestDeploy_GitArgsUsed ensures that any git values provided as flags are used
+// when invoking a remote deployment.
+func TestDeploy_GitArgsUsed(t *testing.T) {
+	root, rm := Mktemp(t)
+	defer rm()
+
+	var (
+		url    = "https://example.com/user/repo"
+		branch = "main"
+		dir    = "function"
+	)
+	// Create a new Function in the temp dir
+	if err := fn.New().Create(fn.Function{Runtime: "go", Root: root}); err != nil {
+		t.Fatal(err)
+	}
+
+	// A Pipelines Provider which will validate the expected values were received
+	pipeliner := mock.NewPipelinesProvider()
+	pipeliner.RunFn = func(f fn.Function) error {
+		if f.Git.URL != url {
+			t.Errorf("Pipeline Provider expected git URL '%v' got '%v'", url, f.Git.URL)
+		}
+		if f.Git.Revision != branch {
+			t.Errorf("Pipeline Provider expected git branch '%v' got '%v'", branch, f.Git.Revision)
+		}
+		if f.Git.ContextDir != dir {
+			t.Errorf("Pipeline Provider expected git dir '%v' got '%v'", url, f.Git.ContextDir)
+		}
+		return nil
+	}
+
+	// Deploy the Function specifying all of the git-related flags and --remote
+	// such that the mock pipelines provider is invoked.
+	cmd := NewDeployCmd(NewClientFactory(func() *fn.Client {
+		return fn.New(fn.WithPipelinesProvider(pipeliner), fn.WithRegistry(TestRegistry))
+	}))
+
+	cmd.SetArgs([]string{"--remote=true", "--git-url=" + url, "--git-branch=" + branch, "--git-dir=" + dir})
+	if err := cmd.Execute(); err != nil {
+		t.Fatal(err)
 	}
 }
+
+// TestDeploy_GitURLBranch ensures that a --git-url which specifies the branch
+// in the URL is equivalent to providing --git-branch
+func TestDeploy_GitURLBranch(t *testing.T) {
+	root, rm := Mktemp(t)
+	defer rm()
+
+	if err := fn.New().Create(fn.Function{Runtime: "go", Root: root}); err != nil {
+		t.Fatal(err)
+	}
+
+	var (
+		url            = "https://example.com/user/repo#branch"
+		expectedUrl    = "https://example.com/user/repo"
+		expectedBranch = "branch"
+	)
+	cmd := NewDeployCmd(NewClientFactory(func() *fn.Client {
+		return fn.New(fn.WithPipelinesProvider(mock.NewPipelinesProvider()), fn.WithRegistry(TestRegistry))
+	}))
+	cmd.SetArgs([]string{"--remote", "--git-url=" + url})
+
+	if err := cmd.Execute(); err != nil {
+		t.Fatal(err)
+	}
+
+	f, err := fn.NewFunction(root)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if f.Git.URL != expectedUrl {
+		t.Errorf("expected git URL '%v' got '%v'", expectedUrl, f.Git.URL)
+	}
+	if f.Git.Revision != expectedBranch {
+		t.Errorf("expected git branch '%v' got '%v'", expectedBranch, f.Git.Revision)
+	}
+}
+
+// TODO: Test that an error is generated if --git-url includes a branch and
+// --git-branch is provided.
+
+// TODO:  --save
```

Comment:
@zroubalik here is the entrypoint for adding the --save flag, which will allow the user to opt-out of having the value of the other flags (`--git-url etc`) written to the func.yaml by default.